### PR TITLE
Add replace label tool for use in GitHub Monorepo

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Add a new tool to the monorepo utils, this adds a `github` tool and specifically a `replace-labels` command that allows you to replace labels in the monorepo.
I wrote this to update the labels for the team Mothra change, and figured it may be handy as a tool.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. In a test repo or WooCommerce repo, create a basic issue with a label that isn't used yet `hacktoberfest` or `team: Mothra`
2. Create a [GH personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) with repository write permissions
3. Run `GITHUB_TOKEN="YOUR_TOKEN" pnpm utils github replace-labels -l "hacktoberfest" -r "type: good first issue"` when `-l` is the label it searches by and `-r` is the replacement label.
4. Make sure it runs correctly and the issues are updated ( revert them back if this change is not intended )
5. Now add `--remove-if-starts-with "team: "` or another prefix of common labels
6. Run the above command again, now making sure you have two labels with the same prefix ( lets say there were two `team` labels and we were looking for the first one.
7. Make sure that the `-l` label is only being removed and not replaced.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: tools/monorepo-utils/src/github/commands/replace-labels/index.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	)
+	.option(
+		'--remove-if-starts-with <removeIfStartsWith>',
+		'Only remove the label if it already contains a label that starts with.'
+	)
+	.option(
+		'-s --source <source>',
+		'Branch to create the release branch from. Default: trunk',
+		'trunk'
+	)
```
Comment (adimoldovan) [inline tools/monorepo-utils/src/github/commands/replace-labels/index.ts]: Copy/paste left-over maybe?

Comment (louwie17) [reply tools/monorepo-utils/src/github/commands/replace-labels/index.ts]: Aaah yes, good catch! Let me remove that.

Comment (louwie17) [reply tools/monorepo-utils/src/github/commands/replace-labels/index.ts]: Removed in: https://github.com/woocommerce/woocommerce/pull/50649/commits/3ff0bbd7485a957562856467ba9f827ce2db55b4

File: tools/monorepo-utils/src/github/commands/replace-labels/index.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			 * Ex: Multiple teams may be assigned to one issue, when replace one team for another
+			 * we did want to keep the team that already exists.
+			 */
+			const containsSimilarLabelAlready =
+				removeIfStartsWith &&
+				labels.find( ( l ) => l.startsWith( removeIfStartsWith ) );
+
+			if ( ! containsSimilarLabelAlready ) {
+				labels.push( replacementLabel );
+			}
```
Comment (psealock) [inline tools/monorepo-utils/src/github/commands/replace-labels/index.ts:91]: Is it worthwhile to limit `replacementLabel` to existing labels? No real preference here, but just thought I'd note.

File: tools/monorepo-utils/src/github/commands/replace-labels/index.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		if ( results.length === 0 ) {
+			Logger.warn( `No issues found by label: "${ label }"` );
+			process.exit( 0 );
+		}
+
+		for ( const issue of results ) {
+			// Get labels by name only and filter out the existing label.
+			const labels = issue.labels
+				.map( ( l ) => ( typeof l === 'string' ? l : l.name ) )
+				.filter( ( l ) => l !== label );
```
Comment (psealock) [inline tools/monorepo-utils/src/github/commands/replace-labels/index.ts]: `getIssuesByLabel` above is not case sensitive, but this check is, so its accurately identifying `Hacktoberfest` with a `-l hacktoberfest` query but the old label isn't removed. We may as well go with the same behavior as Github on this one and make a lowercase comparison.

**Review by psealock — APPROVED** [review]
Nice work! this is working well in my test repo.

I just left a comment and a request to replicate Github's treatment of case sensitivity. Pre-approving

File: tools/monorepo-utils/src/github/commands/replace-labels/index.ts (hunk 2 above)
Comment (louwie17) [reply tools/monorepo-utils/src/github/commands/replace-labels/index.ts:91]: Not a bad idea, I added this as part of this commit: https://github.com/woocommerce/woocommerce/pull/50649/commits/8184dc462b3186244aeb50562f0005f359f53409

File: tools/monorepo-utils/src/github/commands/replace-labels/index.ts (hunk 3 above)
Comment (louwie17) [reply tools/monorepo-utils/src/github/commands/replace-labels/index.ts]: Nice catch, fixed here: https://github.com/woocommerce/woocommerce/pull/50649/commits/8184dc462b3186244aeb50562f0005f359f53409

**Comment by louwie17** [comment]
Thanks for the review @psealock I addressed your comments, are you able to re-review?

**Review by psealock — APPROVED** [review]
Thanks @louwie17, this is working well!! 🚢

**Pull request merged**